### PR TITLE
Case Contact Form Overhaul

### PR DESCRIPTION
**Pull request by thejonroberts**
### What github issue is this PR for, if any?
Resolves #XXXX - Is there an issue for this yet?

Please review/merge these PRs first:
- ~~[ ] #6060~~
- [x] #6061 
- [X] #6057 

They include some of the changes here, but can be merged with no changes to existing behavior. That will make review of this giant PR a little easier.

### What changed, and _why_?
Total overhaul of CaseContact form. Reworked to be in one page. This was accomplished by reducing the wicked wizard steps to one - the details step.

Maintaining the **autosave feature was complicated**! Before, notes were the only thing that could change during an autosave, but now everything is in the form that gets sent during the autosave update. Therefore, `nested_attributes_for` `contact_topic_answers` and `additional_expenses` required a stimulus NestedForm controller that could create and destroy records -- otherwise, autosave would make new records for every save, or try to delete records that had already been removed in the prior update.  This also required creating controllers & routes for those json requests.

### How is this **tested**? (please write tests!) 💖💪
I had to rewrite request and system specs in the process. Request spec was made to be more typical, testing request responses for certain parameters, and not checking page content - I moved those into system specs.  I got rid of the system/create spec and moved those tests to new or edit specs as applicable (I'm not sure how create was different from 'new' in a system spec).  There were also specs asserting things about case show to that spec rather than the form specs, which I moved over to the case show specs.

Also added request and policy specs for the new controller routes, as mentioned.

### Additional Notes
There are a few things from the design requirements that have not been met.  I have pending specs for those (which may need to be removed?). I recommend creating issues and handling them separately. They are:
- Limiting to one answer per ContactTopic, as well as limiting number of topic answers to number of topics
- Showing ContactTopic description in tooltip. This will require swapping the tooltip out on topic select, and I thought there is a lot of javascript logic already.
- There may be some other minor things in the design document.
- I also think it would be good to make an issue for better accessibility. I tried to get the basics out of the way, but one thing that really bothers me is that the form is VERY difficult to tab-navigate through.

### Screenshots please :)
localhost_3000_case_contacts_1105_form_details

### Feelings gif (optional)
_What gif best describes your feeling working on this issue? https://giphy.com/
alt text

### CODE REVIEW

File: config/routes.rb
Comment (thejonroberts) [inline config/routes.rb:198]: I did these as base resources, when they could be member resources, mostly because it was easier. Can change to member resources if needed. Although in a new ticket would be nice.

The constraint for :json was new to me, but I thought it made sense.

File: package.json
Comment (thejonroberts) [inline package.json:44]: I think this was just an older package name, updated to the namespaces package.

File: spec/factories/casa_orgs.rb
Comment (thejonroberts) [inline spec/factories/casa_orgs.rb:23]: arg... merge conflicts that I ignored (-X theirs). dang. will rebase again with more care.

**Comment by bcastillo32** [comment]
wait WOW @thejonroberts thank you so much for jamming on this. You have made my month 😂

**Comment by bcastillo32** [comment]
@thejonroberts thank you for the feedback on the following:

- Requiring ContactType (validation)
- Limiting to one answer per ContactTopic, as well as limiting number of topic answers to number of topics
- Showing ContactTopic description in tooltip. This will require swapping the tooltip out on topic select, and I thought there is a lot of javascript logic already.
- There may be some other minor things in the design document.
- I also think it would be good to make an issue for better accessibility. I tried to get the basics out of the way, but one thing that really bothers me is that the form is VERY difficult to tab-navigate through.

I can create some sep tix for these. As fo this PR @elasticspoon do we need an issue to tie this PR to?

**Comment by bcastillo32** [comment]
> @thejonroberts thank you for the feedback on the following:
> 
> * Requiring ContactType (validation)
> * Limiting to one answer per ContactTopic, as well as limiting number of topic answers to number of topics
> * Showing ContactTopic description in tooltip. This will require swapping the tooltip out on topic select, and I thought there is a lot of javascript logic already.
> * There may be some other minor things in the design document.
> * I also think it would be good to make an issue for better accessibility. I tried to get the basics out of the way, but one thing that really bothers me is that the form is VERY difficult to tab-navigate through.
> 
> I can create some sep tix for these. As fo this PR @elasticspoon do we need an issue to tie this PR to?

@thejonroberts just chatted with Mallory (designer) and confirmed that we should be Requiring ContactType - are you able to do a "hot fix" here or would you like a new ticket?

**Comment by thejonroberts** [comment]
@bcastillo32 I can add the ContactType requirement here.

**Comment by bcastillo32** [comment]
> @bcastillo32 I can add the ContactType requirement here.

Ok great. Thank you so much @thejonroberts !

**Review by elasticspoon — CHANGES_REQUESTED** [review]
Hey! Honestly, this is really good. I really liked the fallback to additional notes when an org has no contact topics set. 

I haven't looked at the code too closely but mainly clicked around a bunch to do some higher level testing:

## Blockers

- delete button for notes does not seem to work?
  - I start a new case contact, add a topic answer, fill it in, reload the page, delete topic answer => is disappears from the page => reload page => its back
  - I see a "you are not authorized' banner at the top of the page (I am admin creating a case contact)
  - logs show `ActionController::RoutingError (No route matches [DELETE] "/"):`
  - should also probably skip ` app/controllers/application_controller.rb:39:in set_active_banner'` for an api only request
- creating a contact topic answer:
  - also tries to `set_active_banner`
  - have this in the logs

```plaintext
ActionController::UnknownFormat (SupervisorsController#index is missing a template for this request format and variant.
request.formats: ["application/json"]
request.variant: []):
```

- I think drafts are broken?
  - logged in as a supervisor, picked a case, new case contact
  - filled in some stuff, did not submit
  - went back to page with all contacts for a case and contact was not there
- I seem to be getting the unauthorized message on random reloads of the contact form


### Nice to have (not blocking)
  - add some client side validation to the required fields
  - mark the driving reimbursement fields are required
  - autosave popup would be nice on other fields as well, just so I know when I am good

**Review by elasticspoon — COMMENTED** [review]
Still working through this. Mainly had some questions. Overall. I think it looks great but obviously its a lot to get through.

File: app/models/case_contact.rb
Comment (thejonroberts) [reply app/models/case_contact.rb:26]: so it goes like:
- case_contacts/new => new record is `status: :started`
- form autosave does not save status during update (within format.json branch)
- html form submit (form_controller#update) sets `status: :details` prior to update (therefore uses the `:active_or_details?` validation)
- if update successful, `#finish_editing` sets `status: :active`
- if update fails, database case_contact will still be `status: :started`

Could just get rid of details status altogether and have contact in either `started` or `active` states. I think it would involve changing these validation ifs and changing the erb template name from `details.html.erb` to `started.html.erb`.  But I didn't want to deal with even more side effects on this PR, I think another ticket for that would be great.

File: app/models/case_contact.rb
Comment (thejonroberts) [reply app/models/case_contact.rb:77]: Will ignore the attrs only if **both** are blank. This is to use a 'fake' contact topic answer as a note.  So we need to allow an 'answer' without a topic_id.  But if that also has no value (no note), then we can ignore it altogether.

File: app/assets/stylesheets/pages/case_contacts_form.scss
Comment (thejonroberts) [reply app/assets/stylesheets/pages/case_contacts_form.scss:1]: Yeah, it's a little rough in the CSS files...  I made issue #6065 related to main.css being separate from the build... 

Roughly half of the rules here are to override elements were styled un-scoped in other files, or had a frustrating default.

Some cleanup & sensible base variables & styles would go a long way. Then sprinkles of bootstrap classes in the templates, and mostly layout concerns in view / component css might be possible?  It be a big job.

File: app/controllers/additional_expenses_controller.rb
Comment (thejonroberts) [reply app/controllers/additional_expenses_controller.rb:23]: hmm.. rails generated controllers don't worry about it by default... I think the only way a would fail is if there is a  `dependent: :restrict_with_exception` on a model relation, or some bad `dependent: destroy` chain. I think we are safe here since this isn't a 'parent' of any other records.

File: app/policies/additional_expense_policy.rb
Comment (thejonroberts) [reply app/policies/additional_expense_policy.rb:6]: Honestly am always confused about the array syntax here. Forget every time.  But I had to change this, so hold that thought... the relation doesn't work through CasaCases because CaseContacts don't have a casa_case_id until they are active.

Should have updates tomorrow.

File: app/policies/additional_expense_policy.rb
Comment (thejonroberts) [reply app/policies/additional_expense_policy.rb:18]: You mean something like
```suggestion
      user.case_contacts.exists?(record.case_contact_id)
```
or something else?

File: app/models/additional_expense.rb
Comment (thejonroberts) [reply app/models/additional_expense.rb:1]: Yeah, less than ideal. But allowing only for 0 amounts means it shouldn't matter for accounting. I don't see a way around it without a different autosave approach?

File: app/assets/stylesheets/pages/case_contacts_form.scss
Comment (elasticspoon) [reply app/assets/stylesheets/pages/case_contacts_form.scss:1]: >  It'd be a big job.

For sure

File: app/policies/additional_expense_policy.rb
Comment (elasticspoon) [reply app/policies/additional_expense_policy.rb:18]: Uhhhh. Im not sure tbh if that works. But `include?` will pop us into ruby land and traverse the whole array to find that contact. I'm like 99% sure we could do it fully in the db. But if there isn't a super simple obvious way don't waste your time on it.

File: app/models/case_contact.rb
Comment (elasticspoon) [reply app/models/case_contact.rb:26]: Gotcha that makes sense.

**Comment by FireLemons** [comment]
QA'd this tonight as part of office hours.

Some things I noticed that were

Hitting the add note button triggers the autosave and saves a blank note.

The autosave seems to only be triggered by the add note button and any inputs made to the `textarea` of the notes row but it saves the state of all inputs which I think is unexpected behavior. I recommend only saving the note text or triggering the autosave for all form changes to make it more intuitive.

Add note button:
image

**Comment by thejonroberts** [comment]
> ## Blockers
> * delete button for notes does not seem to work?
>   * I start a new case contact, add a topic answer, fill it in, reload the page, delete topic answer => is disappears from the page => reload page => its back
>   * I see a "you are not authorized' banner at the top of the page (I am admin creating a case contact)
>   * logs show `ActionController::RoutingError (No route matches [DELETE] "/"):`
>   * should also probably skip ` app/controllers/application_controller.rb:39:in set_active_banner'` for an api only request
> * creating a contact topic answer:
>   * also tries to `set_active_banner`

Really glad you found these issues, I definitely should have tried things out as admin and supervisor roles. Addressed the two main problems here:
- setting banner for json requests - return unless an html request in set banner.
- ContactTopicAnswer Policy was not working because of a nil casa_org relation through CaseContact (case_contact.casa_org_id is nil until contact is 'activated'). I added a 'backup' casa_org relation to go through `creator_id` instead (creator_casa_org). Topic Answers and Expenses Policies will use the new relation if the 'regular' casa_org is not present. I also changed the CaseContactPolicy to use this, and changed it to not permit cross-organization actions (couldn't help myself).

>  * I think drafts are broken?
>   * logged in as a supervisor, picked a case, new case contact
>   * filled in some stuff, did not submit
>   * went back to page with all contacts for a case and contact was not there
> * I seem to be getting the unauthorized message on random reloads of the contact form

So, this is related to above - we don't set the `CaseContact.casa_case_id` until the contact is activated. CasaCase show page won't know about them because the relation isn't there yet.  Unless I'm missing something, I don't think drafts were listed there before these changes? They will be shown as drafts in CaseContacts index, but not CasaCase show. Could make a good ticket to keep `draft_case_ids.first` -> `casa_case_id` in sync during draft process? 

If I'm wrong and drafts were there before, I'm stumped at the moment.

> ### Nice to have (not blocking)
> * add some client side validation to the required fields 

Separate ticket, please. I could easily add `required` attributes to inputs, but I think the feedback from server validation is better than browser built in. Using the AR messages translations and such. A more complicated js approach is too much here.

> * mark the driving reimbursement fields are required 
- DONE

> * autosave popup would be nice on other fields as well, just so I know when I am good

As @FireLemons pointed out, the form only auto-saves during Note text input currently.  The thought was to maintain previous functionality. Autosave happened for notes because they could be lengthy and **very** painful if you are forced to start over.
- I could add a `data-action="input->autosave#save"` on every single input to fire off an autosave on every change
  - Where to show notifications? Top of every section? No design for these, my want to think on it. 
- OR we could just autosave every X seconds 
  - will save a lot of template actions, and make fewer server requests... unless someone walks away from their computer, and it just sends the same update over and over...
- Maybe a combo, autosave on any input, but stretch out the debounce time to like 30 seconds or a minute?

If we have a decision on these things, or get some more feedback, I'm happy to do those here, but I think that as long as there's no loss of functionality, this is a good place to stop feature-wise, and then iterate based on user feedback?

File: app/policies/additional_expense_policy.rb
Comment (thejonroberts) [reply app/policies/additional_expense_policy.rb]: Changed to the [exists?](https://edgeapi.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-exists-3F) approach.  Good eye! I wasn't thinking about queries when I should have been.

**Review by thejonroberts — COMMENTED** [review]
Alright, this is in a place that I am very happy with, thanks for the feedback so far, very helpful.

I just spent a lot of time fleshing out Policy, PolicyScope and their specs for CaseContacts, AdditionalExpenses, and ContactTopicAnswers. Definitely take a look at those / their specs.

I do ask for no new features here unless absolutely necessary, as mentioned above. But missing requirements, bugginess, confusing/smelly code, missing specs etc... are very welcome!

I am sorry this PR ballooned into such a behemoth!

File: app/controllers/application_controller.rb
Comment (thejonroberts) [inline app/controllers/application_controller.rb:120]: Is this comment clear?

Avoids this:
```ruby
if @additional_expense.save
    render json: @additional_expense.as_json, status: :created
else
    render json: @additional_expense.errors.as_json, status: :unprocessable_entity
end
```
Needing to be (something like) this:
```ruby
respond_to do |format|
  if @additional_expense.save
    format.json { render json: @additional_expense.as_json, status: :created }
  else
    format.json { render json: @additional_expense.errors.as_json, status: :unprocessable_entity }
  end
end
```

File: spec/support/fill_in_case_contact_fields.rb
Comment (thejonroberts) [reply spec/support/fill_in_case_contact_fields.rb:1]: Yeah, definitely feels a bit indirect now. I think describing actions in the test make more sense now that it's all on one page.  Maybe just leaving a "fill required fields"/"happy_path" type helper to DRY things up, and everything non-happy path spelled out in the actual specs. I may take a stab at that. But I don't think it hurts to leave it as is.

File: app/controllers/application_controller.rb
Comment (thejonroberts) [inline app/controllers/application_controller.rb:168]: Above the `ActionController::UnknownFormat` for the new controllers, as well as any other unhandled format. Can add formats here (xml, csv, etc...) as needed.

**Comment by thejonroberts** [comment]
@elasticspoon , @FireLemons Regarding the autosave questions and my previous comments:

> - I could add a `data-action="input->autosave#save"` on every single input to fire off an autosave on every change
>   - Where to show notifications? Top of every section? No design for these, my want to think on it. 
> - OR we could just autosave every X seconds 
>   - will save a lot of template actions, and make fewer server requests... unless someone walks away from their computer, and it just sends the same update over and over...
> - Maybe a combo, autosave on any input, but stretch out the debounce time to like 30 seconds or a minute?

I think a decent and easy enough solution for now would be to add autosave actions for Additional Expense Description inputs. The expenses may also be difficult to remember/re-enter if you get logged out. I'll add that. I think everything else in the form is easy enough to re-do if you get logged out, and with notes/expenses at the bottom of the form, you will hit an autosave after you have filled them in (typically).

Does that sound like a good approach to y'all?

**Comment by thejonroberts** [comment]
I have added the change I mentioned above. **Expense description input now autosaves**. Moved notifications to bottom of the Note & Expenses section 'cards'. Also now hides the notifications after a delay -- much easier to tell when the updates happen vs. leaving "Saved!" notification up indefinitely.

**Comment by elasticspoon** [comment]
I have a long weekend so I'll have time to look over this in the next couple days

**Comment by MalloryPriceDesign** [comment]
Hi @thejonroberts, this looks awesome! I'm really excited to see this single-page form coming together. I've just finished reviewing your work for design, so a lot of my feedback will be around aligning the style and UX with our [Figma file](https://www.figma.com/design/QHOE6wHBESpsbxQmlnAZ2L/Case-Contact-Reporting-Process?node-id=2287-4858&node-type=frame). I've split this into parts that should be addressed immediately (if possible) and things that could be split into future tickets. Let me know what you think!


## To address in this PR:
- [ ] **Date auto-populating:**
        In the past, some users have complained about submitting forms where data was automatically input for them, such as date. In the Figma file, I was hoping to address this by leaving the date field unpopulated until a user clicks on it. This initial click will still populate it with today's date, as well as open the calendar window. 

- [ ] **Days since the last contact:**
        In the past, we've included an italicized "About 1 week ago" since the last contact note for each Contact Type. Can we add this back? Users found this helpful when filling out the form. 
        Screen Shot 2024-10-12 at 3 24 37 PM. 

- [ ] **Contact Medium options:**
        Can you render these options vertically instead of horizontally? This should increase clarity and accessibility.

- [ ] **Including One Discussion Note:**
        When the page is initialized, we should include one discussion topic and note. Users report wanting to be able to write their notes down quickly, so having a textbox already available should save them a click. Also, can you confirm that these Discussion Topics are being pulled from the organization's configuration? We'll be using these values to map the notes to their corresponding section in the generated court report.

- [ ] **Limit to one entry per topic in Notes:**
        If it's not too hard, can we limit the number of entries for each topic to one? I'm assuming this should also help our handling of these topics on the backend when generating reports so we don't have to stitch multiple entries per topic together.

- [ ] **Reimbursement Visibility**
        Can you confirm that the whole reimbursement card is only visible if the user's administrator has enabled reimbursement? Additionally, I cannot see the "Other Expense Amount" and "Other Expense Details" inputs in this section.


## Additional feedback

**Overall UI Feedback**
1. Consistent horizontal padding - should match Figma
    1. Between "Relevant Case(s)" and "Contact Date"
    2. Between "Hours" and "Minutes"
    3. Between "Discussion Topics" and "Discussion Notes"
    4. Between "Total miles driven" and "Current address"
    5. Padding between the checkboxes and the circle checkboxes and the label should all be the same
2. Consistent vertical padding - should match Figma
    1. Between each "Discussion Notes" box (vertical padding)
    2. Padding between labels and their associated input box (ie, "Relevant Case(s)" and its input box)
    3. Align padding above "Recrd New Case Contact" with the Figma file
4. Align the copy to match the Figma file in the text boxes. 
    1. For example, update alternative text in the "Relevant Case(s)" input box, from "Select or search" to "Select or search for case(s)"
5. Color for contact type Group Names should be  #5D657B
6. Align sizes for textboxes, they should all be 20x20, see Figma. Similar note for the circles in the multiple choice questions (ie, "Contact Medium" options)
7. The options for "Contact Medium" should be vertical
8. All Labels should be Title Case, ie "Total Miles Driven"
9. Dummy text to be added where it's missing


**Notes Section**
1. The "Discussion Topic" should not be populated by default, with the "Select a discussion topic" alt text showing.
2. Are these topics hard-coded? They should be pulled from the admin control panel
3. The form should initialize with one Note visible. The topic should not be selected.
4. The alt text should be present in the Topic and Notes boxes if nothing is selected or inputted.
5. Color of the Discussion Topic and Discussion Notes should match the format of headers for other labels, ie "Relevant Case(s)," "Contact Dates"

**Reimbursement Section**
1. Are we including the "Other Expense Amount" and "Other Expense Details" input boxes? If not, let's make a ticket for them
2. Are we maintaining functionality where if an admin has not enabled reimbursement for this org, this whole box isn't viewable to the volunteer
3. The color of labels should match the format of other labels (similar to the above feedback in the Notes section)
4. Is it possible to make the height of the "total miles driven" and "current address" input boxes the same?
5. Update copy to "Mailing Address for Reimbursement Check"


Please let me know if you have any questions about my feedback. Great work so far!

**Comment by elasticspoon** [comment]
> @elasticspoon , @FireLemons Regarding the autosave questions and my previous comments:
> 
> > * I could add a `data-action="input->autosave#save"` on every single input to fire off an autosave on every change
> >   
> >   * Where to show notifications? Top of every section? No design for these, my want to think on it.
> > * OR we could just autosave every X seconds
> >   
> >   * will save a lot of template actions, and make fewer server requests... unless someone walks away from their computer, and it just sends the same update over and over...
> > * Maybe a combo, autosave on any input, but stretch out the debounce time to like 30 seconds or a minute?
> 
> I think a decent and easy enough solution for now would be to add autosave actions for Additional Expense Description inputs. The expenses may also be difficult to remember/re-enter if you get logged out. I'll add that. I think everything else in the form is easy enough to re-do if you get logged out, and with notes/expenses at the bottom of the form, you will hit an autosave after you have filled them in (typically).
> 
> Does that sound like a good approach to y'all?

Yea. I'm good with that as an initial approach and we can see in the future if we need to change that.

File: spec/policies/case_contact_policy_spec.rb
Comment (elasticspoon) [reply spec/policies/case_contact_policy_spec.rb:74]: No they should not. The notes have potentially private information.

**Review by elasticspoon — CHANGES_REQUESTED** [review]
Most of my stuff is just small comments but the drafts on the case details page (/casa_cases/:case_slug) don't appear, which is actually blocking.


https://github.com/user-attachments/assets/6af72b99-2cd2-40f3-8b2a-c793a0cf7dc1

File: spec/policies/case_contact_policy_spec.rb
Comment (thejonroberts) [reply spec/policies/case_contact_policy_spec.rb]: We have to set it in the controller, anyway. Sanity check to ensure we don't allow creating something for the wrong creator/org.

File: app/controllers/application_controller.rb
Comment (thejonroberts) [reply app/controllers/application_controller.rb:120]: That worked, but was just kind of a strange pattern to me -- I felt like this was a more rails-y way to do it/other people could read & understand it better. I would look in the routes file AFTER the controller, anyway.

File: spec/policies/contact_topic_answer_policy_spec.rb
Comment (thejonroberts) [reply spec/policies/contact_topic_answer_policy_spec.rb:1]: Yeah, although theses are pretty basic, too. We could probably just defer to the parent CaseContact for these policies? Might be something to discuss along with CaseContactPolicy questions.

File: spec/system/case_contacts/new_spec.rb
Comment (thejonroberts) [reply spec/system/case_contacts/new_spec.rb]: yeah, I just removed the back button, since it is not in the design & now equivalent to the browser back button!

**Comment by thejonroberts** [comment]
@MalloryPriceDesign thanks so much for the detailed feedback! I have reworked the form quite a bit and it is much better now! Notes:

> ## To address in this PR:
> * [x]  **Date auto-populating:**

I have made the date input blank on load. Note that there is some weird behavior in Safari browsers - The date input's value is null, but it is displayed as today's date. So user may think it is set, but value is null and will error if form is submitted as is. This could be confusing, so I added basic browser validation to require an input before submit... I think that feedback could probably be improved, depends on the browser implementation.

I also could not get the CSS for input placeholder text color to work consistently for all inputs, I think due to our bootstrap setup. These issues would make good improvement tickets?

> * [x]  **Days since the last contact:**
> * [x]  **Contact Medium options:**
> * [x]  **Including One Discussion Note:**
>   When the page is initialized, we should include one discussion topic and note. Users report wanting to be able to write their notes down quickly, so having a textbox already available should save them a click. Also, can you confirm that these Discussion Topics are being pulled from the organization's configuration? We'll be using these values to map the notes to their corresponding section in the generated court report.

- There is one note shown on load now, unless editing an active contact that had no answers, or organization has no contact topics.
- These topics are from Organization's contact topics. If there are none available, a disclaimer is shown (as previously), and an 'Additional Notes' field is rendered for contact.notes column.

> * [ ]  **Limit to one entry per topic in Notes:**
>   If it's not too hard, can we limit the number of entries for each topic to one? I'm assuming this should also help our handling of these topics on the backend when generating reports so we don't have to stitch multiple entries per topic together.

- I haven't done this yet but will try. I thought of a solution that is likely workable - disabling options that are currently selected... adding and removing completely from list of options is more complicated. Going to attempt it now.
- I think case contact report will be affected. (uses `index_by(&:contact_topic_id)`: last topic answer will be shown, not others)

> * [x]  **Reimbursement Visibility**
>   Can you confirm that the whole reimbursement card is only visible if the user's administrator has enabled reimbursement? Additionally, I cannot see the "Other Expense Amount" and "Other Expense Details" inputs in this section.

Yes, the entire Reimbursement section card is not rendered unless organization driving reimbursement is enabled. 

> 1. Are we including the "Other Expense Amount" and "Other Expense Details" input boxes? If not, let's make a ticket for them
> 2. Are we maintaining functionality where if an admin has not enabled reimbursement for this org, this whole box isn't viewable to the volunteer

Expenses are behind another feature flag, so they are not shown unless Driving Reimbursement & Expense reimbursement  are both enabled. Note that this means creator must request driving reimbursement to request Other Expenses... So you can have no reimbursement, driving reimbursement only, or driving & expense reimbursement. But not only Expense reimbursement.

> ## Additional feedback
> **Overall UI Feedback**

Reworked classes and CSS to be much more consistent with design. Pretty close I think, the Figma files were extremely helpful, thank you.

> **Notes Section**
> 1. The "Discussion Topic" should not be populated by default, with the "Select a discussion topic" alt text showing.

This is a little more complex than I want to deal with right now. Because of a database constraint on `contact_topic_id` (not null), we need a topic id when a new contact topic answer (Discussion Note) is added to the form. I think this could be implemented if the database constraint is removed, which would then allow us to require the Topic ID conditionally.  Alternatively, could do some (overly complicated imo) javascript manipulations? Regardless, I think it is better left for future work.

> Reimbursement Section
> 4. Is it possible to make the height of the "total miles driven" and "current address" input boxes the same?

The only way to do this (easily) means that it can only be a one line entry, as opposed to a more traditional address format, is that okay? Will wait to make the change until that is confirmed.

**Comment by thejonroberts** [comment]
@MalloryPriceDesign - ready for re-review, see above, edited some notes I made earlier. (may need to be deployed to the environment you were looking at, I'm not sure how that was setup)

@elasticspoon - ready for re-review, with the caveat that I have not addressed the CasaCase show page draft issue, as I don't think that was a feature prior to this work, unless I'm missing something.

**Comment by MalloryPriceDesign** [comment]
> > Reimbursement Section
> > 4. Is it possible to make the height of the "total miles driven" and "current address" input boxes the same?

> The only way to do this (easily) means that it can only be a one line entry, as opposed to a more traditional address format, is that okay? Will wait to make the change until that is confirmed.

Let's keep this as-is, I think the users will probably expect the traditional address format instead of a one-line format.


> @MalloryPriceDesign - ready for re-review, see above, edited some notes I made earlier. (may need to be deployed to the environment you were looking at, I'm not sure how that was setup)

That's great! I used Codespaces to review previously. However, after running `git pull` and building the app again in my previous Codespace, I got an error. I restarted with a fresh Codespace and am getting the same error:
> This page isn’t working
supreme-disco-rj9q4grx54jfppxg-3000.app.github.dev is currently unable to handle this request.
HTTP ERROR 502

I'll take a look again later.

**Comment by elasticspoon** [comment]
> > Most of my stuff is just small comments but the drafts on the case details page (/casa_cases/:case_slug) don't appear, which is actually blocking.
> 
> @elasticspoon, that cannot be done currently in the app! I tried it out on main branch to confirm.
> 
> We don't set `CaseContact.casa_case_id` until the final submission of the contact. Selected case(s) are in `draft_case_id` column only. The CasaCase `has_many :case_contacts` relationship is not there yet, so Case show view is unaware of the drafts. You can only see draft contacts from the CaseContact index (/case_contacts).

You are right. My flabbers are ghasted. I totally thought that was something that could be done.

**Review by elasticspoon — APPROVED** [review]
@thejonroberts This is pretty tremendous work man. You are a saint to doing all this work and being so thorough about it. I don't see any blocking issues. I'm sure something will come up but I'm fine to send this as is. 

I'll let @MalloryPriceDesign chime in if she has any further input.

**Comment by MalloryPriceDesign** [comment]
Looks amazing! I rebuilt my Codespace container and it's working now!

This looks good! The only change I would suggest is to enable the calendar dropdown to display on the initial click into the date box. Currently, the user has to go all the way to the right calendar icon to click to open the calendar. Is that something that could be easily changed? If not, I don't think this should block merging and can always be addressed in another ticket.

Screen Shot 2024-10-20 at 9 30 22 PM

**Comment by thejonroberts** [comment]
Thank you, Mallory...

> This looks good! The only change I would suggest is to enable the calendar dropdown to display on the initial click into the date box. Currently, the user has to go all the way to the right calendar icon to click to open the calendar. Is that something that could be easily changed? If not, I don't think this should block merging and can always be addressed in another ticket.

This is browser dependent -- Safari opens date picker on click for example. I think any changes to browser input behavior should be separate to make sure time can be spent ensuring cross-browser behaviors and accessibility considerations are maintained, a lot of gotchas can pop up!

**Comment by bcastillo32** [comment]
thank you everyone for the work on this! This is amazing and feedback on the codespace was very positive!

**Pull request merged**
